ortools/linear_solver/xpress_interface.cc

```diff
@@ -1931,9 +1999,104 @@ void XpressInterface::AddSolutionHintToOptimizer() {
+  auto callback_with_context = static_cast<const MPCallbackWithXpressContext*>(cbdata);
+  CHECK(callback_with_context != nullptr);
+  CHECK(callback_with_context->context != nullptr);
+  CHECK(callback_with_context->callback != nullptr);
+  if (callback_with_context->context->UpdateFromXpressState(cbprob)) {
+    // Only run user's callback if the context has changed
+    callback_with_context->callback->RunCallback(
+        callback_with_context->context);
+  }
+}
+
+XpressMPCallbackContext::XpressMPCallbackContext(XPRSprob& mLp, bool mip,
+                                                 bool maximize, int num_vars)
+    : mLp_(mLp),
+      mip_(mip),
+      maximize_(maximize),
+      num_vars_(num_vars),
+      variable_values_(num_vars_, XPRS_NAN),
+      objective_value_(XPRS_NAN),
+      best_objective_bound_(XPRS_NAN),
+      num_nodes_(0),
+      num_user_sols_(0) {}
+
+MPCallbackEvent XpressMPCallbackContext::Event() {
+  return MPCallbackEvent::kMipSolution; // this is the only supported callback for now
+}
+
+bool XpressMPCallbackContext::CanQueryVariableValues() {
+  return !std::isnan(objective_value_);
+}
+
+double XpressMPCallbackContext::VariableValue(const MPVariable* variable) {
+  return variable_values_[variable->index()];
+}
+
+double XpressMPCallbackContext::SuggestSolution(
+    const absl::flat_hash_map<const MPVariable*, double>& solution) {
+  // Currently the XPRESS API does not handle clearing out previous hints
+  const std::size_t len = solution.size();
+  if (len == 0) {
+    // hint is empty, do nothing
+    return XPRS_NAN;
+  }
+  unique_ptr<int[]> colind(new int[len]);
+  unique_ptr<double[]> val(new double[len]);
+  int i = 0;
+  for (const auto& [var, value] : solution) {
+    colind[i] = var->index();
+    val[i] = value;
+    ++i;
   }
+  // use a new name so new solutions don't override old ones
+  std::string name = "USER_CB_HINT_" + std::to_string(++num_user_sols_);
+  XPRSaddhint(mLp_, len, val.get(), colind.get(), name.c_str());
+
+  // XPRESS doesn't guarantee when it will test the suggested solution
+  // So we return the last known objective value but with no guarantee that it
+  // is computed on the suggested solution
+  return objective_value_;
+}
+
+bool XpressMPCallbackContext::UpdateFromXpressState(XPRSprob cbprob) {
+  if (mip_ && num_vars_ > 0) {
+    double new_objective_value;
+    CHECK_STATUS(
+        XPRSgetdblattrib(cbprob, XPRS_MIPOBJVAL, &new_objective_value));
+    bool improved = (std::isnan(objective_value_)) ||
+                    (maximize_ && new_objective_value > objective_value_) ||
+                    (!maximize_ && new_objective_value < objective_value_);
+    if (!improved) {
+      // If objective function value has not improved, do nothing.
+      // This can actually happen since XPRESS can callback multiple times with
+      // the same solution. Exiting here helps avoid unnecessary computations.
+      return false;
```

Comment:
Why is a change in the dual bound not considered an improvement? If you want to invoke callbacks only for changes in the primal solution you may want to register an [intsol](xprsaddcbintsol) callback?

Reply:
Hi Daniel, thank you for your precious feedback, I will try to take into account your comments in order to improve the code.

Regarding this question, in fact we only want the callback to be called only when a new MIP solution is found (see operations_research::MPCallbackEvent::kMipSolution). I must have missed XPRSaddcbintsol when searching for the proper CB function, I will use it instead.
However, it seems to me that in XPRSaddcbintsol, there is no garantee that the new solution is better than the last best one? I wonder if it's acceptable from the user's perspective, otherwise we still have to check the objective function before notifying them. What do you think?

Reply:
Thinking about it, from a user perspective, I don't think that it's a big problem if every solution you get is not better than the last one. The user can filter out some solutions depending on their needs. I will make this code more simple.

Reply:
In a multi-threaded solve you cannot guarantee monotonicity between callbacks anyway: two threads may invoke the callback simultaneously with different solutions. This is just inherent to multi-threading.
So better to not assume this "always improving" feature.